core/trino-main/src/main/java/io/trino/operator/GroupByIdBlock.java

```diff
@@ -13,199 +13,186 @@
+ * full (111...111) for a standard encoding and empty (000...000) for a rle.
+ * This way the rle block will always return the first position from the underlying array.
+ * <p>
+ * The implementation guarantees that every fetch of a single value will not
+ * result in a virtual call and, as a result, may be safely inlined by the compiler.
+ * <p>
+ * The only limitation is that the block cannot be dictionary encoded, but this encoding
+ * is not used within group by logic.
+ */
 public class GroupByIdBlock
         implements Block
 {
     private static final int INSTANCE_SIZE = ClassLayout.parseClass(GroupByIdBlock.class).instanceSize();
+    private static final int SIZE_OF_ELEMENT = Long.BYTES;
+
+    private static final int MASK_STANDARD = -1;
+    private static final int MASK_RLE = 0;
 
     private final long groupCount;
-    private final Block block;
+    private final int positionCount;
+    private final long[] groups;
+    private final int mask;
 
-    public GroupByIdBlock(long groupCount, Block block)
+    public static GroupByIdBlock standard(long groupCount, long[] groupIds)
     {
-        requireNonNull(block, "block is null");
-        this.groupCount = groupCount;
-        this.block = block;
+        return new GroupByIdBlock(groupCount, groupIds.length, groupIds, MASK_STANDARD);
     }
 
-    public long getGroupCount()
+    public static GroupByIdBlock rle(long groupCount, long value, int length)
     {
-        return groupCount;
+        return new GroupByIdBlock(groupCount, length, new long[] {value}, MASK_RLE);
     }
 
-    public long getGroupId(int position)
+    public static GroupByIdBlock ofBlock(long groupCount, Block block)
     {
-        return BIGINT.getLong(block, position);
+        if (block instanceof GroupByIdBlock) {
+            return (GroupByIdBlock) block;
+        }
+        long[] groups = new long[block.getPositionCount()];
+        for (int i = 0; i < block.getPositionCount(); i++) {
+            groups[i] = block.getLong(i, 0);
+        }
+        return GroupByIdBlock.standard(groupCount, groups);
     }
 
-    @Override
-    public Block getRegion(int positionOffset, int length)
+    private GroupByIdBlock(long groupCount, int positionCount, long[] groups, int mask)
     {
-        return block.getRegion(positionOffset, length);
+        this.groupCount = groupCount;
+        this.positionCount = positionCount;
+        this.groups = requireNonNull(groups, "groups is null");
+        this.mask = mask;
     }
 
-    @Override
-    public long getRegionSizeInBytes(int positionOffset, int length)
+    public long getGroupCount()
     {
-        return block.getRegionSizeInBytes(positionOffset, length);
+        return groupCount;
     }
 
-    @Override
-    public OptionalInt fixedSizeInBytesPerPosition()
+    public long getGroupId(int position)
     {
-        return block.fixedSizeInBytesPerPosition();
+        checkReadablePosition(position);
+        return groups[position & mask];
```

Comment:
The jit does have some optimizations for booleans... did you try `groups[rle ? 0 : position]`.   Not really important, just curious

Reply:
We did some experimentation with `sum += flag ? 1: 0` a while back and it usually inlines as a cast to byte without any branches. That wasn't, however, always the case, so the explicit mask approach seems more certain.